content/articles/memoization-and-dp-in-python-beginners-guide/index.md

```diff
@@ -0,0 +1,122 @@
+In this article, we are going to talk about a dynamic programming technique called memoization and we are going to use it to solve an optimization problem.
+
+###Table of contents
+[Prerequisites](#prerequisites)
+
+[Memoization](#memoization)
+
+[The 0-1 knapsack problem](#knapsack)
+
+[Python Solution to the 0-1 knapsack problem](#implementation)
+
+[How to apply memoization to the solution](#solution)
+
+###prerequisites<a name='prerequisites'></a>
+1. A basic understanding of data structures and algorithms
+2. Basic knowledge of Python
+3. A basic understanding of Recursion
+### what is memoization? <a name='memoization'></a>
+memoization is the process of saving the results of a function to a cache in order to improve its performance.
+It is the technique where we ensure that the function does not run more than once for the same inputs; instead, we return a cached result to reduce the running time.
+It is similar to tabulation except that while tabulation uses the bottom-up approach, memoization uses a top-down approach. This means that when approaching a problem with tabulation, we save the results starting with smaller, lower values going up whereas in memoization we start from the larger values going down.
+In the fibonacci example given in the tutorial linked above we can see that we started from values 2 up towards the value passed on to the function.
+```python
+for i in range(2, n):
+   term = array[i - 1] + array[i - 2]
+   array.append(term)
+   # return the nth term of the array
+return array[n - 1]
+```
+The memoized version of this is going to look like this:
+```python
+def mem_fib(n, array):
+    '''
+    n is the fib number,
+    array is an array of n+1 numbers
+    '''
+    if n<=1:
+        array[n]=n
+    if array[n] is None:
+        array[n] = mem_fib(n-1, array)+mem_fib(n-2, array)
+    return array[n]
+```
+In the memoized version you can see that we started from the values n-1 and n-2 going downwards.
+Now that you know what memoization is, let's use it to solve a common programming problem.
+###The 0-1 Knapsack problem<a name='knapsack'></a>
+The 0-1 knapsack problem is an optimization problem where you are given a set of items, their weights and values, and a knapsack. Your task is to determine the combination of items that are going to give maximum value without reaching the weight limit of your knapsack.
+For every item in the list, we can either include it in the knapsack or discard it(This is where the 0-1 name comes from).
+In the coding implementation of this problem, you are given two arrays, one is the weights array and the other is the values array. You are also given a weight W which is the maximum weight the knapsack can carry.
+From the definition alone, we can see that this problem can be solved using Dynamic Programming. This is because it has overlapping subproblems that when solved lead towards solving the whole problem.
```

Comment:
`From the definition alone, we can see that this problem can be solved using Dynamic Programming.`

Beginners might not be aware of DP or other programming stuff. You need to break them down.

Explain what DP is, why we use it? How is it related with DP? Why not brute force?